src/xpath.c

```diff
@@ -6728,107 +6811,111 @@ eval_literal(struct lyxp_expr *exp, uint16_t *exp_idx, struct lyxp_set *set)
 }
 
 /**
- * @brief Check and parse a predicate following a leaf-list to extract its value.
- * On success all the used tokens are skipped.
+ * @brief Try to compile list or leaf-list predicate in the known format to be used for hash-based instance search.
  *
- * @param[in] exp Parsed XPath expression.
- * @param[in] exp_idx Position in the expression @p exp.
- * @param[out] val Leaf-list value on success.
- * @param[out] val_len Length of @p val.
- * @return LY_ERR
+ * @param[in] exp Full parsed XPath expression.
+ * @param[in,out] exp_idx Index in @p exp at the beginning of the predicate, is updated on success.
+ * @param[in] scnode Found schema node as context for the predicate.
+ * @param[in] format Format of any prefixes in key names/values.
+ * @param[out] predicates Parsed predicates.
+ * @param[out] pred_type Type of @p predicates.
+ * @return LY_SUCCESS on success,
+ * @return LY_ERR on any error.
  */
 static LY_ERR
-eval_name_test_parse_leaflist_predicate(struct lyxp_expr *exp, uint16_t *exp_idx, const char **val, uint16_t *val_len)
+eval_name_test_try_compile_predicates(struct lyxp_expr *exp, uint16_t *exp_idx, const struct lysc_node *scnode,
+                                      LYD_FORMAT format, struct ly_path_predicate **predicates, enum ly_path_pred_type *pred_type)
 {
-    uint16_t cur_idx;
-
-    cur_idx = *exp_idx;
-
-    /* '[' */
-    LY_CHECK_RET(lyxp_check_token(NULL, exp, *exp_idx, LYXP_TOKEN_BRACK1));
-    ++cur_idx;
+    LY_ERR ret = LY_SUCCESS;
+    uint16_t key_count, e_idx, pred_idx = 0;
+    const struct lysc_node *key;
+    size_t pred_len;
+    int prev_lo;
+    struct lyxp_expr *exp2 = NULL;
 
-    /* '.' */
-    LY_CHECK_RET(lyxp_check_token(NULL, exp, *exp_idx, LYXP_TOKEN_DOT));
-    ++cur_idx;
+    assert(scnode->nodetype & (LYS_LIST | LYS_LEAFLIST));
 
-    /* '=' */
-    LY_CHECK_RET(lyxp_check_token(NULL, exp, *exp_idx, LYXP_TOKEN_OPER_EQUAL));
-    ++cur_idx;
+    if (scnode->nodetype == LYS_LIST) {
```

Comment:
Not sure, is it necessary something besides the keys count check here, since there is the ly_path_parse_predicate() later? Maybe even the keys count check can be done using the parsed expression structure.

Reply:
Right, first some background for this code. The idea here is to check that these predicates follow a certain simple pattern (accepted for `ly_path`) of key-value pairs. However, it covers also a situation when there are these predicates but are followed by more predicates, which is fine for XPath and in that case we want to parse the known predicates, find the desired instance using hashes, and then apply the following predicates. For this reason this function must already know how many keys there are so that it knows how many predicates are to be parsed by `ly_path_parse_predicate`. Yes, that function also checks the key count again (but also key names and everything) so that is performed twice but I do not think that is a problem at all. Fixing this would require some changes for those `ly_path` functions only for this one use-case and I would consider that a worse solution that the current one.